### PR TITLE
Add warning before overwrite file

### LINKED ISSUE #997: Warn when about to overwrite file
### Checklist

- [X] I have checked the [issue page](https://github.com/lay295/TwitchDownloader/issues?q=is%3Aissue) for duplicates
- [X] I am running the latest version ([download here](https://github.com/lay295/TwitchDownloader/releases/latest))

### Write your feature request here

It would be nice if the program would warn you if it's about to overwrite files when you enqueue render. I just had a few stream videos overwritten (and the Twitch past broadcasts are expired) because I didn't realize the JSON file was the same name as the video file name. Thanks!

### PR DESCRIPTION
**Pull request by atomsbaza**
- Resolves #997 

TODO:
- Warning message and ReadLine input to confirm before overwriting the file

Feedback is welcomed.

### CODE REVIEW

**Comment by ScrubN** [comment]
Thank you for the PR, however there are several issues:
1. This will only work for the CLI. As it stands, this would completely break the Windows GUI
2. Continuing on `y`, else failing is poor design in my opinion. The user should be explicitly required to confirm or reject, otherwise the program should ask again.
3. How will this work for 3rd party libraries that wrap TDCLI? They cannot reasonably work around this.
4. How will this work with status disabled? (`--log-level None`)
5. What about chat updater and chat renderer?

**Comment by atomsbaza** [comment]
> Thank you for the PR, however there are several issues:
> 
> 1. This will only work for the CLI. As it stands, this would completely break the Windows GUI
> 2. Continuing on `y`, else failing is poor design in my opinion. The user should be explicitly required to confirm or reject, otherwise the program should ask again.
> 3. How will this work for 3rd party libraries that wrap TDCLI? They cannot reasonably work around this.
> 4. How will this work with status disabled? (`--log-level None`)
> 5. What about chat updater and chat renderer?

Thank you for your feedback. To answer question

1. For Windows GUI I think it isn't needed because Windows already asks to confirm before saving the file but I'm not sure that will cover all necessary because I'm using macOS so I will look for a solution again.
2. I'm gonna improve it as you ask.
3. I'm gonna looking ways to make it work. 
4. I will add more logs that can be used from ITaskLogger.
5. I will add a warning for chat updater and chat rederer

**Comment by ScrubN** [comment]
> 1. For Windows GUI I think it isn't needed because Windows already asks to confirm before saving the file but I'm not sure that will cover all necessary because I'm using macOS so I will look for a solution again.

The Windows save file dialog does not delete files when a user clicks "yes". Because of that, your changes break the Windows GUI completely due to scanning `Console.In`.
rider64_wmqsy6BIuZ (1)

**Comment by atomsbaza** [comment]
> > 1. For Windows GUI I think it isn't needed because Windows already asks to confirm before saving the file but I'm not sure that will cover all necessary because I'm using macOS so I will look for a solution again.
> 
> The Windows save file dialog does not delete files when a user clicks "yes". Because of that, your changes break the Windows GUI completely due to scanning `Console.In`. rider64_wmqsy6BIuZ (1) [ rider64_wmqsy6BIuZ (1) ](https://private-user-images.githubusercontent.com/72096833/322053458-34e963cb-35ce-43ac-a6a4-fed412c6c507.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI5NTQzODcsIm5iZiI6MTcxMjk1NDA4NywicGF0aCI6Ii83MjA5NjgzMy8zMjIwNTM0NTgtMzRlOTYzY2ItMzVjZS00M2FjLWE2YTQtZmVkNDEyYzZjNTA3LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDEyVDIwMzQ0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiYjg1ZjgyYjgzOWI3NTQ0Yzc3MjMyMjkwMTRiZjMzODQwZDRmNjVkNTViZmY3NjkzNTcwMGY2MDFlMjFiY2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Kf5_oDStX1YZkrulnZvnci3zp6ywEeEeoEXNUJ259KY) [ ](https://private-user-images.githubusercontent.com/72096833/322053458-34e963cb-35ce-43ac-a6a4-fed412c6c507.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTI5NTQzODcsIm5iZiI6MTcxMjk1NDA4NywicGF0aCI6Ii83MjA5NjgzMy8zMjIwNTM0NTgtMzRlOTYzY2ItMzVjZS00M2FjLWE2YTQtZmVkNDEyYzZjNTA3LmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDEyVDIwMzQ0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJiYjg1ZjgyYjgzOWI3NTQ0Yzc3MjMyMjkwMTRiZjMzODQwZDRmNjVkNTViZmY3NjkzNTcwMGY2MDFlMjFiY2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Kf5_oDStX1YZkrulnZvnci3zp6ywEeEeoEXNUJ259KY)




I see what happen now.
Thank you for your feedback will work in-progress

**Comment by atomsbaza** [comment]
@ScrubN could you please explain more about the status disabled (--log-level None)

**Comment by ScrubN** [comment]
When `--log-level None` is passed, nothing is written to stdout except for uncaught exceptions.

**Comment by atomsbaza** [comment]
Thanks, @ScrubN Could you please review this PR after I complete my changes?

File: TwitchDownloaderCore/ChatDownloader.cs [hunk 1]
```diff
@@ -244,13 +244,36 @@ private static List<Comment> ConvertComments(CommentVideo video, ChatFormat form
...
+                        _progress.SetStatusWarning("Operation aborted. File not overwritten.");
+                        return;
+                    }
+                    else
+                    {
+                        _progress.SetStatusWarning("Invalid response. Please enter 'confirm' to overwrite or 'reject' to cancel.");
+                    }
+                }
+            }
+
```
Comment (ScrubN) [inline TwitchDownloaderCore/ChatDownloader.cs:276]: DRY - Don't repeat yourself. You have the exact same block of code in 5 different files.

File: TwitchDownloaderCore/ChatDownloader.cs [hunk 2]
```diff
@@ -244,13 +244,36 @@ private static List<Comment> ConvertComments(CommentVideo video, ChatFormat form
...
+                    }
+                    else if (response == "reject")
+                    {
+                        _progress.SetStatusWarning("Operation aborted. File not overwritten.");
+                        return;
+                    }
+                    else
+                    {
+                        _progress.SetStatusWarning("Invalid response. Please enter 'confirm' to overwrite or 'reject' to cancel.");
+                    }
```
Comment (ScrubN) [inline TwitchDownloaderCore/ChatDownloader.cs:273]: `Confirm` and `Reject` aren't very common and may be difficult for some ESL users. `[Y]es`/`[N]o` is more common and already used in CacheHandler.

File: TwitchDownloaderCore/ChatDownloader.cs [hunk 3]
```diff
@@ -244,13 +244,36 @@ private static List<Comment> ConvertComments(CommentVideo video, ChatFormat form
...
             {
                 throw new NullReferenceException("Null or empty video/clip ID");
             }
 
+            if(allowOverwrite) 
+            {
+                while (File.Exists(downloadOptions.Filename))
+                {
+                    _progress.SetStatusWarning("File already exists. Do you want to overwrite it? (Confirm/Reject)");
+                    var response = Console.ReadLine()?.Trim().ToLower();
```
Comment (ScrubN) [inline TwitchDownloaderCore/ChatDownloader.cs:259]: As mentioned earlier, this breaks scripts who cannot write to stdin.

File: TwitchDownloaderCore/ChatDownloader.cs [hunk 4]
```diff
@@ -244,13 +244,36 @@ private static List<Comment> ConvertComments(CommentVideo video, ChatFormat form
             return returnList;
         }
 
-        public async Task DownloadAsync(CancellationToken cancellationToken)
+        public async Task DownloadAsync(CancellationToken cancellationToken, bool allowOverwrite = true)
```
Comment (ScrubN) [inline TwitchDownloaderCore/ChatDownloader.cs:247]: Why is this a method parameter when everything else is part of `...Options`.

Comment (ScrubN) [inline TwitchDownloaderCore/ChatDownloader.cs:247]: This breaks WPF tasks.

File: TwitchDownloaderCore/ChatDownloader.cs (hunk 3 above)
Comment (ScrubN) [inline TwitchDownloaderCore/ChatDownloader.cs:259]: Why are we doing CLI specific operations in the general purpose core library?

**Review by ScrubN — CHANGES_REQUESTED** [review]
You missed `TsMerge`.

**Pull request closed without merging**